From: Hans Verkuil <hverkuil-cisco@xs4all.nl>
Date: Thu, 2 May 2024 13:38:41 +0200
Subject: [PATCH] media: cec: cec-adap: always cancel work in cec_transmit_msg_fh
Message-ID: <ee47a5df845643f3eee272b951535d7b43547074.1714649924.git.hverkuil-cisco@xs4all.nl>
Patch-mainline: Submitted, linux-media ML
References: CVE-2024-23848 bsc#1219104

Do not check for !data->completed, just always call
cancel_delayed_work_sync(). This fixes a small race condition.

Signed-off-by: Hans Verkuil <hverkuil-cisco@xs4all.nl>
Reported-by: Yang, Chenyuan <cy54@illinois.edu>
Fixes: 490d84f6d73c ("media: cec: forgot to cancel delayed work")
Closes: https://lore.kernel.org/linux-media/PH7PR11MB57688E64ADE4FE82E658D86DA09EA@PH7PR11MB5768.namprd11.prod.outlook.com/
Acked-by: Takashi Iwai <tiwai@suse.de>

---
 drivers/media/cec/cec-adap.c |    1 +
 1 file changed, 1 insertion(+)

--- a/drivers/media/cec/cec-adap.c
+++ b/drivers/media/cec/cec-adap.c
@@ -769,6 +769,7 @@ int cec_transmit_msg_fh(struct cec_adapt
 	mutex_unlock(&adap->lock);
 	res = wait_for_completion_killable_timeout(&data->c,
 						   msecs_to_jiffies(timeout));
+	cancel_delayed_work_sync(&data->work);
 	mutex_lock(&adap->lock);
 
 	if (data->completed) {
